### DISCUSSION TITLE
Basic Plotting (Spectrum and Time-series)

### DISCUSSION THREAD

**Discussion by JuliDi** [FutureSDR/FutureSDR#108] [Q&A] [answered]
Hey,

in the past couple of months I've lost track of this project, so maybe there is already some working examples that I am unaware of.

For starters, I would like to do some basic signal analysis, i.e., plot the spectrum of the signal and plot the signal over time. This is something I know GnuRadio is capable of (and it works okayish).
The setup would roughly look like this: `SoapySource -> FFT ---> Plot` and `Soapy Source ----> Plot`. In practice there would be some more blocks to this, of course.

Is there an example for how to do this with FutureSDR, e.g. with python using matplotlib or so? If not, how could I achieve this?

I presume, there is no integrated plotting facilities in FutureSDR yet, since afaik there is not yet any great plotting backends for Rust. Some time ago I got the spectrum example working via WASM, but it was a pain in the neck with the USB drivers for web I am not familiar with all the web frameworks used. So it was also a bit overkill if I didn't need it to run on the web.

**Comment by bastibl** [comment, marked as answer]
There is the spectrum example, which should provide roughly what you are suggesting. (Start the flowgraph (`cargo run --bin cpu --features cpu`) and go to http://127.0.0.1:1337/spectrum3d.html or http://127.0.0.1:1337/api/fg/0/ for some json description of the flowgraph. There is also the new futuresdr-remote crate which should simplify interaction with a running flowgraph. https://docs.rs/futuresdr-remote/latest/futuresdr_remote/)

Instead of a web GUI, you could also use a Python script to read the data from the websocket:
https://twitter.com/bastibl/status/1625528632587304962
https://github.com/bastibl/maia-sdr/blob/main/maia-httpd/util/spectrum_client.py

Running everything inside the browser is possible, but pretty tricky. It's at the moment only a proof-of-concept.

(I just found some bugs, when testing the example. Please try (for now) with the `main` branch at github.com/bastibl/futuresdr)

**Reply by JuliDi** [reply]
Great, thanks for all the input. I will give it a try.
Not aiming for running everything in the browser anyway, that easily ends up getting too complicated and difficult to debug.

**Reply by JuliDi** [reply]
It turned out to work really nicely with the websockets example and I managed to get live plotting with plotters in Rust.
Once it's in a usable state, I will share a link to the repo :)

**Comment by metasim** [comment]
Related: https://github.com/FutureSDR/FutureSDR/pull/139
